packages/zero-pg/src/custom.ts

```diff
@@ -75,22 +78,173 @@ export class TransactionImpl<S extends Schema, TWrappedTransaction>
+      );
+      const tx = this[dbTxSymbol];
+      await tx.query(stmt.text, stmt.values);
+    },
+    async upsert(this: WithHiddenTx, value) {
+      const targetedColumns = origAndServerNamesFor(Object.keys(value), schema);
+      const primaryKeyColumns = origAndServerNamesFor(
+        schema.primaryKey,
+        schema,
+      );
+      const stmt = formatPg(
+        sql`INSERT INTO ${sql.ident(serverName(schema))} (${sql.join(
+          targetedColumns.map(([, serverName]) => sql.ident(serverName)),
+          ',',
+        )}) VALUES (${sql.join(
+          Object.values(value).map(v => sql.value(v)),
+          ', ',
+        )}) ON CONFLICT (${sql.join(
+          primaryKeyColumns.map(([, serverName]) => sql.ident(serverName)),
+          ', ',
+        )}) DO UPDATE SET ${sql.join(
+          Object.entries(value).map(
+            ([col, val]) =>
+              sql`${sql.ident(
+                schema.columns[col].serverName ?? col,
+              )} = ${sql.value(val)}`,
+          ),
+          ', ',
+        )}`,
+      );
+      const tx = this[dbTxSymbol];
+      await tx.query(stmt.text, stmt.values);
+    },
+    async update(this: WithHiddenTx, value) {
+      const targetedColumns = origAndServerNamesFor(Object.keys(value), schema);
+      const stmt = formatPg(
+        sql`UPDATE ${sql.ident(serverName(schema))} SET ${sql.join(
+          targetedColumns.map(
+            ([origName, serverName]) =>
+              sql`${sql.ident(serverName)} = ${sql.value(value[origName])}`,
+          ),
+          ', ',
+        )} WHERE ${primaryKeyClause(schema, value)}`,
+      );
+      const tx = this[dbTxSymbol];
+      await tx.query(stmt.text, stmt.values);
+    },
+    async delete(this: WithHiddenTx, value) {
+      const stmt = formatPg(
+        sql`DELETE FROM ${sql.ident(
+          serverName(schema),
+        )} WHERE ${primaryKeyClause(schema, value)}`,
+      );
+      const tx = this[dbTxSymbol];
+      await tx.query(stmt.text, stmt.values);
+    },
+  };
+}
+
+export function makeSchemaQuery<S extends Schema>(
+  _schema: Schema,
+): (dbTransaction: DBTransaction<unknown>) => SchemaQuery<S> {
+  return (_dbTransaction: DBTransaction<unknown>) => {
+    throw new Error('Not implemented');
+  };
+}
+
+function serverName(x: {name: string; serverName?: string | undefined}) {
+  return x.serverName ?? x.name;
+}
+
+function primaryKeyClause(schema: TableSchema, row: Record<string, unknown>) {
+  const primaryKey = origAndServerNamesFor(schema.primaryKey, schema);
+  return sql`${sql.join(
+    primaryKey.map(
+      ([origName, serverName]) =>
+        sql`${sql.ident(serverName)} = ${sql.value(row[origName])}`,
+    ),
+    ' AND ',
+  )}`;
+}
+
+function origAndServerNamesFor(
+  originalNames: readonly string[],
+  schema: TableSchema,
+): [origName: string, serverName: string][] {
+  return originalNames.map(name => {
+    const col = schema.columns[name];
+    return [name, col.serverName ?? name] as const;
```

Comment:
we should make `serverName` non-optional in the underlying structure to remove these checks. It can remain optional in the builder interface but later filled in by `createSchema`.

Reply:
we should also validate the schema.ts that exists on the server matches the user's pg db when `createHandler` is invoked.

Related: https://github.com/rocicorp/mono/pull/3783